### PR TITLE
Add cloudbees-folder dependency

### PR DESCRIPTION
**Pull request by jglick**
Counterpart to https://github.com/jenkinsci/cloudbees-folder-plugin/pull/33, and would have to be released simultaneously to minimize the chance of loss of functionality. (I do not know of a way to _force_ users who upgrade `cloudbees-folder` to also upgrade `matrix-auth`, alas.)

@reviewbybees esp. @stephenc

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -12,13 +12,12 @@
     <description>Offers matrix-based security authorization strategies (global and per-project).</description>
     <url>http://wiki.jenkins-ci.org/display/JENKINS/Matrix+Authorization+Strategy+Plugin</url>
     <properties>
-      <findbugs-maven-plugin.version>3.0.1</findbugs-maven-plugin.version>
-      <findbugs.failOnError>true</findbugs.failOnError>
+      <jenkins.version>1.625.3</jenkins.version> <!-- must have a65d007 -->
```
Comment (jglick) [inline pom.xml]: https://github.com/jenkinsci/jenkins/commit/a65d007 by @Vlatombe. Not needed for _running_ the plugin, but needed for _testing_ it: with the `cloudbees-folder` dep, we must also have the `credentials` dep, and then you get a plugin cycle without this core patch. Cf. [JENKINS-28942](https://issues.jenkins-ci.org/browse/JENKINS-28942).

File: src/main/java/com/cloudbees/hudson/plugins/folder/properties/AuthorizationMatrixProperty.java [hunk 2]
```diff
@@ -0,0 +1,283 @@
...
+import java.util.Map.Entry;
+import java.util.Set;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import jenkins.model.Jenkins;
+
+/**
+ * Holds ACL for {@link ProjectMatrixAuthorizationStrategy}.
+ */
+public class AuthorizationMatrixProperty extends AbstractFolderProperty<AbstractFolder<?>> {
```
Comment (jglick) [inline src/main/java/com/cloudbees/hudson/plugins/folder/properties/AuthorizationMatrixProperty.java]: I think a move without edits.

File: src/main/java/hudson/security/AuthorizationMatrixProperty.java [hunk 3]
```diff
@@ -173,7 +173,7 @@ protected void add(Permission p, String sid) {
 		@Override
 		public boolean isApplicable(Class<? extends Job> jobType) {
             // only applicable when ProjectMatrixAuthorizationStrategy is in charge
-            return Jenkins.getInstance().getAuthorizationStrategy() instanceof ProjectMatrixAuthorizationStrategy;
+            return Jenkins.getActiveInstance().getAuthorizationStrategy() instanceof ProjectMatrixAuthorizationStrategy;
```
Comment (jglick) [inline src/main/java/hudson/security/AuthorizationMatrixProperty.java]: forced by core upgrade

File: src/main/java/hudson/security/ProjectMatrixAuthorizationStrategy.java [hunk 4]
```diff
@@ -70,7 +71,7 @@ public ACL getACL(Job<?,?> project) {
 
     private static ACL inheritingACL(final ACL parent, final ACL child) {
         if (parent instanceof SidACL && child instanceof SidACL) {
-            return ((SidACL) parent).newInheritingACL((SidACL) child);
+            return ((SidACL) child).newInheritingACL((SidACL) parent);
```
Comment (jglick) [inline src/main/java/hudson/security/ProjectMatrixAuthorizationStrategy.java]: Amends a change made in #8. @stephenc I believe you got the receiver and argument backward when you did that.

File: src/main/java/hudson/security/ProjectMatrixAuthorizationStrategy.java [hunk 5]
```diff
@@ -83,13 +84,21 @@ public boolean hasPermission(Authentication a, Permission permission) {
     public ACL getACL(ItemGroup g) {
         if (g instanceof Item) {
             Item item = (Item) g;
-            return (SidACL)item.getACL();
+            return item.getACL();
```
Comment (jglick) [inline src/main/java/hudson/security/ProjectMatrixAuthorizationStrategy.java]: Forgotten in #8; without this I get a (different) CCE from the `permissions` test in https://github.com/jenkinsci/workflow-plugin/pull/334.

File: src/main/java/hudson/security/ProjectMatrixAuthorizationStrategy.java [hunk 6]
```diff
@@ -83,13 +84,21 @@ public boolean hasPermission(Authentication a, Permission permission) {
...
     @Override
     public ACL getACL(AbstractItem item) {
+        if (Jenkins.getActiveInstance().getPlugin("cloudbees-folder") != null) { // optional dependency
+            if (item instanceof AbstractFolder) {
+                com.cloudbees.hudson.plugins.folder.properties.AuthorizationMatrixProperty p = (com.cloudbees.hudson.plugins.folder.properties.AuthorizationMatrixProperty) ((AbstractFolder) item).getProperties().get(com.cloudbees.hudson.plugins.folder.properties.AuthorizationMatrixProperty.class);
+                if (p != null) {
+                    return inheritingACL(getACL(item.getParent()), p.getACL());
+                }
+            }
+        }
```
Comment (jglick) [inline src/main/java/hudson/security/ProjectMatrixAuthorizationStrategy.java]: The meat of the change; makes the downstream test pass.

File: src/main/resources/index.jelly [hunk 7]
```diff
@@ -1,3 +1,4 @@
+<?jelly escape-by-default='true'?>
```
Comment (jglick) [inline src/main/resources/index.jelly]: forced by new parent POM

File: src/test/java/com/cloudbees/hudson/plugins/folder/properties/AuthorizationMatrixPropertyTest.java [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+import hudson.security.HudsonPrivateSecurityRealm;
+import hudson.security.ProjectMatrixAuthorizationStrategy;
+import java.util.concurrent.Callable;
+import org.acegisecurity.AccessDeniedException;
+import static org.junit.Assert.*;
+import org.junit.Rule;
+import org.junit.Test;
+import org.jvnet.hudson.test.JenkinsRule;
+
+public class AuthorizationMatrixPropertyTest {
```
Comment (jglick) [inline src/test/java/com/cloudbees/hudson/plugins/folder/properties/AuthorizationMatrixPropertyTest.java]: Moved without edits as far as I recall.

**Comment by jglick** [comment]
Forgot to deploy dependency snapshot.

**Comment by andresrc** [comment]
:bee:

**Comment by stephenc** [comment]
:bee:

**Pull request merged**